virttest/qemu_vm.py

```diff
@@ -2544,6 +2553,7 @@ def create(self, name=None, params=None, root_dir=None,
         if params is not None:
             self.params = params
             self.devices = None     # Representation changed
+            self._reprobe_capabilities()
```

Comment:
At this point we are not yet creating the machine, therefor you should not update the capabilities (in case of a failure we'd have different caps than the caps of the running VM).

My problem with all those `_caps` is that it is reinventing the wheel. We have the `self.*` attributes that are initialized on `__init__` and store the current vm's values. Later we needed to define non-persistent state, which is the `self.devices` and it was developed and treated so it performs well. Later spice-team added the `self.spice_options` only to find-out that it is not easy to make it to work properly so I had to modify it so it works the same way as `self.devices`. Now the `_caps` got developed with the same mistakes and many workarounds are being added.

I'd recommend removing the `_caps` completely and use `self.devices` to store the caps. Then we could use the already existing methods to re-create it.

Reply:
Also I'd suggest to avoid the optimization with Flags and simply used properties/dicts. Yes, there is a minor overhead, but it's not in any critical path and the overhead is minimal. On the other hand it confuses and is harder to use (first develop, then measure and only when necessary optimize)

Reply:
.. so instead of `self._caps` one would use `"blockdev" in self.devices.caps` and in `make_create_command` one would use `vm.devices.caps` to get the capabilities (which should IMO be set or dict).

Reply:
... note that keeping the Capabilities and Flags is fine by me (although I dislike it). My problem is with the unnecessary API to make it work, which has already been done for `self.devices`. So simply by moving it there you can remove all the recreate/create/... stuff and simply use it.

Reply:
> At this point we are not yet creating the machine, therefor you should not update the capabilities (in case of a failure we'd have different caps than the caps of the running VM).
> 
> My problem with all those `_caps` is that it is reinventing the wheel. We have the `self.*` attributes that are initialized on `__init__` and store the current vm's values. Later we needed to define non-persistent state, which is the `self.devices` and it was developed and treated so it performs well. Later spice-team added the `self.spice_options` only to find-out that it is not easy to make it to work properly so I had to modify it so it works the same way as `self.devices`. Now the `_caps` got developed with the same mistakes and many workarounds are being added.
> 
> I'd recommend removing the `_caps` completely and use `self.devices` to store the caps. Then we could use the already existing methods to re-create it.

@ldoktor Inspired by libvirt, there are two types of capabilities: domain capabilities and hypervisor capabilities. It is clear that domain capabilities should be stored in `vm` whereas hypervisor capabilities could be stored in either `vm` or `vm.devices`, but I guess you won't like such a complicated implementation.

Here we are talking about domain capabilities. If it is stored in `vm.devices`, since `params` does not participate during `DevContainer.__init__`, how to respect users decision? Let's take blockdev as an example, hypervisor would report the capability if qemu >= 2.12.0, however, user may want it to not be exposed by the given domain, then he can suppress it by setting the related param. If there is no `params`, how to know that? I would not suggest implementing that by introducing additional argument to the `DevContainer.__init__`, because it will mess up the interface when we have so many demands later and you are passing domain caps to `DevContainer` essentially.

Reply:
> Also I'd suggest to avoid the optimization with Flags and simply used properties/dicts. Yes, there is a minor overhead, but it's not in any critical path and the overhead is minimal. On the other hand it confuses and is harder to use (first develop, then measure and only when necessary optimize)

Using `Flags` can help us detect misspelling than using dicts. I have nothing against using properties, it looks good to me as well.

Reply:
Well the hypervisor capabilities can be stored outside `vm` (even in borg classes throughout the whole avocado-vt execution) but unless there are too many of them (or they are expensive to query) it IMO makes sense to bundle them with the domain capabilities.

Storing them in `vm.devices` would resolve the re-creating as this class is designed for it. I see 2 ways to implement this (sorted according to my preferences):

1. Do auto-detection and always report the capabilities that are supported and use the params on usage (therefor `image_define_by_params` would look whether to use forced value or capabilities).
2. Pass whole params (or just the individual set) to `DevContainer.__init__` and it'd report either the forced or detected capabilities based on the params.

(1) is IMO cleaner as you'd always get the real capability, therefor potential tests might check whether it's supported and then override it from within the test

(2) is simpler as the test would not need to apply the params to get forced or real value.

At this point I only know of `-blockdev` flag and that one is treated only in Avocado-vt (`qemu_vm` and `qemu_devices`), therefor passing the params there shouldn't matter. In the future it might become pain (like migrate params) or benefit (in case we willingly want to check supported vs. forced values).

I can imagine a combination that might easy the borders, but let's start with a basic implementation first... As for `Flags` vs. `properties` vs. `dict`, all are fine by me. Ideally I'd use `properties` as they are the most natural for me, then dict as they are very simple. I don't like much flags (provided they are implemented as enums) as they are more C-like, than python (at least in my eyes), but as long as they work I'm fine with them :-)

Reply:
> Storing them in `vm.devices` would resolve the re-creating as this class is designed for it. I see 2 ways to implement this (sorted according to my preferences):
> 
>     1. Do auto-detection and always report the capabilities that are supported and use the params on usage (therefor `image_define_by_params` would look whether to use forced value or capabilities).

Well, that is what I mean by hypervisor capabilities, to my understanding the main difference between the two types of capabilities is domain capabilities can be affected by users willingness while hypervisor capabilities shouldn't.

In this case, in order to use forced value, we have to introduce additional argument or attribute to `images_define_by_variables` as `image_define_by_params` cannot directly pass params to the former.

> 
>     2. Pass whole params (or just the individual set) to `DevContainer.__init__` and it'd report either the forced or detected capabilities based on the params.

That would work, but I'd like to know why the `params` isn't used previously, is there some reason for that?

And here is another question, by storing `caps` in `vm.devices`, how do users use this attribute (domain capabilities) before VMs are "created"? In other words, the question could be: consider using the same states definition as libvirt (https://wiki.libvirt.org/page/VM_lifecycle#States_that_a_guest_domain_can_be_in), what is the state of a VT VM before it calls `create` and which states should `caps` be presented?

To me, it makes sense that users shouldn't use `caps` before VMs were created only if VMs were in `undefined` state at that moment.

> 
> 
> (1) is IMO cleaner as you'd always get the real capability, therefor potential tests might check whether it's supported and then override it from within the test
> 
> (2) is simpler as the test would not need to apply the params to get forced or real value.
> 
> At this point I only know of `-blockdev` flag and that one is treated only in Avocado-vt (`qemu_vm` and `qemu_devices`), therefor passing the params there shouldn't matter. In the future it might become pain (like migrate params) or benefit (in case we willingly want to check supported vs. forced values).

So, in that case I'd suggest using hypervisor capabilities for checking supported and domain capabilities for checking forced values.

> 
> I can imagine a combination that might easy the borders, but let's start with a basic implementation first... As for `Flags` vs. `properties` vs. `dict`, all are fine by me. Ideally I'd use `properties` as they are the most natural for me, then dict as they are very simple. I don't like much flags (provided they are implemented as enums) as they are more C-like, than python (at least in my eyes), but as long as they work I'm fine with them :-)

OK, got it, thanks.